feat: Add probe mode in debugger

xxuejie opened the pull request:
This change introduces a new probe mode in ckb-debugger, which provides uprobes exposing ckb-vm's full register state. We have to use 3 uprobes, since uprobe has limits on the arguments supported.

XuJiandong commented on ckb-debugger/src/main.rs:
```diff
@@ -394,5 +394,85 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
+                        machine.add_cycles(cycles).map(|_| inst)
+                    })
+                    .and_then(|inst| {
+                        let ra = machine.registers()[registers::RA];
```
It's better to use one probe to expose a memory pointer to registers at startup:
https://github.com/nervosnetwork/ckb-vm/pull/272/files#diff-7de6cee2dbf7962fc5e95c3cedd3d5b9f27b720544c9941663f1b329d704bba6R79-R81

And let bpftrace locate register if needed.

xxuejie replied:
Done, thanks for the suggestion.

xxuejie commented:
This is now updated to expose memory in probes as well. Since it's hard to look into actual memory content when using SparseMemory, we also change the code to use FlatMemory when `probes` feature is enabled.

We have to use raw git version of depended crates now, since the required PRs supporting this has not been included in proper releases.

xxuejie commented:
Waiting for https://github.com/nervosnetwork/ckb/pull/3859 to land in a proper CKB release, then we will rebase this PR and (hopefully) merge it to tip.

contrun commented on ckb-debugger/src/main.rs:
```diff
@@ -394,5 +401,75 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
+                        probe!(ckb_vm, execute_inst, pc, cycles, inst, regs, memory);
+                        let r = execute(inst, &mut machine);
+                        let cycles = machine.cycles();
+                        probe!(
```
In https://github.com/nervosnetwork/ckb-standalone-debugger/pull/71, I tried to add two additional user-friendly probe for syscalls and functions. I also added helper scripts two probe function calls and returns in https://github.com/contrun/ckb-vm-bpf-toolkit/tree/uretprobe/. In order for typical ckb developers to use these probes, we need to document them. Once we documented them, we need to keep them more or less stable. I think `regs` and `memory` are two quite useful common arguments for all these probes. Can we make them as the first and second arguments for all probes? We need also change the helper scripts in https://github.com/xxuejie/ckb-vm-bpf-toolkit.

xxuejie replied:
I wasn't sure about the 2 new probes for syscalls and functions. To me, we can reuse the generic probe, and does syscall / function call detection within BPF scripts. Adding new probes here might be an endless path: today we decided syscalls are important, what if one day we noticed that jumps are important as well? Or instructions touching SP is important? I do suggest a different path where we build more features within BPF code, not adding more probes.

I don't have any preferences on argument orders, tho, we can discuss them here, once there is a finalized order, I can make the change.

xxuejie committed:
feat: Add probe mode in debugger

This change introduces a new probe mode in ckb-debugger, which
provides uprobes exposing ckb-vm's full register state. We have to
use 3 uprobes, since uprobe has limits on the arguments
supported.
